### ISSUE TITLE
Shadowing of Pytorch's torch.utils.data.DataLoader is incomplete

### ISSUE THREAD

**Issue by antorsae** [fastai/fastai#2638]
`fastai2.DataLoader` is almost API-compatible with `torch.utils.data.DataLoader`, however `sampler` is ignored altogether:

```
dataset,weights = list(range(10)), list(range(10))
pytorch_dl = torch.utils.data.DataLoader(
    dataset=dataset,sampler=torch.utils.data.WeightedRandomSampler(weights,len(weights)),batch_size=20)
print(list(pytorch_dl))
fastai2_dl = DataLoader(
    dataset=dataset,sampler=torch.utils.data.WeightedRandomSampler(weights,len(weights)),batch_size=20)
print(list(fastai2_dl))
```
```
[tensor([9, 3, 4, 7, 7, 7, 4, 7, 7, 4])]
[tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])]
```

As far as I can understand, fastai2 uses `get_idxs` internally for sampling strategies, but gets ignored e.g.     
or is assumed to be provided in other places e.g. https://github.com/fastai/fastai/issues/2637

**Comment by jph00** [comment]
I'm planning to thoroughly look at sampler next week - in particular to see if we can avoid the slowness on Windows.

**Comment by antorsae** [comment]
Im going to submit a PR about https://github.com/fastai/fastai/issues/2637 and indirectly touches some of the assumptions/conventions made by `DataLoader`. Some comments/questions:

1. `get_idxs` is executed in the context of `_FakeLoader` b/c is executed at the beginning of `sample()`. This means that each worker will execute `get_idxs`. Is that intentional? Wouldn't it be better to have something like:

```
   def sample(self):
        return (b for i,b in enumerate(self._idxs) if i//(self.bs or 1)%self.nw==self.offs)

    def __iter__(self):
        self.randomize()
        self.before_iter()
        self._idxs=self.get_idxs()
        for b in _loaders[self.fake_l.num_workers==0](self.fake_l):
            if self.device is not None: b = to_device(b, self.device)
            yield self.after_batch(b)
        self.after_iter()
        if hasattr(self, 'it'): delattr(self, 'it')
```
so that `get_idxs` gets executed only once in the context of the main process?

2. What are the minimum assumptions about `DataLoader`? Pytorch's `DataLoader` has a clear set of assumptions, but with `DataLoader` I'm not sure. The most notorious one is `n`, one example:

```
    class RepeatedDL(TfmdDL):
    def __init__(self,dl,repeats,num_workers=None):
        repeats=max(repeats,1)
        if num_workers is None: num_workers = min(16, defaults.cpus)
        super().__init__(dataset=dl.dataset,bs=dl.bs,num_workers=num_workers,pin_memory=dl.pin_memory,timeout=dl.timeout,
                        shuffle=dl.shuffle,drop_last=dl.drop_last,indexed=dl.indexed,n=dl.n,device=dl.device,
                        after_batch=dl.after_batch)
        store_attr(self, 'dl,repeats')

    def get_idxs(self): return np.repeat(self.dl.get_idxs(),self.repeats).tolist()
    def __len__(self): return len(self.dl) * self.repeats
```
The purpose of `RepeatedDL` is take an existing `DataLoader` and just repeat items, e.g.  0 0 0 1 1 1 2 2 2 (if repeats=3). This may be useful for TTA so you can also see w/ `show_batch` how different augmentations look like. `RepeatedDL` purposedly decides to keep `dl.n` the same and just modifies `__len__` and `get_idxs`. `get_idxs` returns more items than `self.n`; and the assumption I made is to treat  `self.n` as if it was private to the original `dl` making no assumptions about its use; but other people may make different assumptions about relation between `self.n` and `get_idxs` and `__len__`.

**Comment by avn3r** [comment]
Has this been addressed?

By far the most wanted feature for FastAI for me is to be able to simply use pytorch samplers. It will make the DataBlock API complete, DataBlock support splitter, labeling function it should also support sampler function or pytorch sampler class. That will open up many unique sampling techniques for problems that require triplet loss, self-learning problems, etc...

**Comment by jph00** [comment]
DataBlock is a separate issue. DataLoader already lets you use custom samplers.

--
  Jeremy Howard
  j@howard.fm

**Comment by muellerzr** [comment]
@jph00 do we know if this can be closed yet or not? :)

**Comment by jph00** [comment]
Yup can be - `get_idxs` is used to change sampling method
